### PR TITLE
Add support for FS-DAX kind in ctl_create_tier_kind_from_env

### PR DESCRIPTION
**Pull request by michalbiesek**
Required:
- [x] #595 

~~TODO:~~
- [x] tests for real FS-DAX space

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/596)

### CODE REVIEW

**Review by PatKamin — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/596)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @michalbiesek)

**Review by bratpiorka — COMMENTED** [review]
do we know why ls -s crashes? is it because of some bug in tiering?

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, 1 of 1 files at r5, 2 of 2 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/596#-:-MWdY8gaAzifl3x8xzDz:b-e4l55k)* status: all files reviewed, 3 unresolved discussions (waiting on @michalbiesek)

___
*[test/tiering_tests.py, line 235 at r6](https://reviewable.io/reviews/memkind/memkind/596#-MWdXccX4EeDUE66ycRH:-MWdXccY06nhB2KH1AAi:bxilstl) ([raw file](https://github.com/memkind/memkind/blob/66d23d75162ad78465a71995fc537ffcfbaffbfc/test/tiering_tests.py#L235)):*
> ```Python
>             "MEMKIND_MEM_TIERING_CONFIG=FS_DAX:/tmp/:" +
>             pmem_size + ":1," + self.default_policy,
>             log_level="2", validate_retcode=False)
> ```

with this change retcode should be validated
___
*[test/tiering_tests.py, line 252 at r6](https://reviewable.io/reviews/memkind/memkind/596#-MWdXlNvD3dfaZ7rLi6n:-MWdXlNvD3dfaZ7rLi6o:b3kqx) ([raw file](https://github.com/memkind/memkind/blob/66d23d75162ad78465a71995fc537ffcfbaffbfc/test/tiering_tests.py#L252)):*
> ```Python
>             "MEMKIND_MEM_TIERING_CONFIG=FS_DAX:/tmp/:" +
>             pmem_size + ":1," + self.default_policy,
>             log_level="2", validate_retcode=False)
> ```

.
___
*[tiering/ctl.c, line 20 at r5](https://reviewable.io/reviews/memkind/memkind/596#-MWdWzEl16p7fLbm6veI:-MWdWzEl16p7fLbm6veJ:b1xekei) ([raw file](https://github.com/memkind/memkind/blob/3ef3aad49ad6ff2dae8a4e5d50185d0daa3cccf1/tiering/ctl.c#L20)):*
> ```C
> static struct memtier_tier *current_tier;
> 
> typedef struct fs_dax_registy {
> ```

registy -> registry

File: tiering/ctl.c [hunk 1]
```diff
@@ -17,13 +17,42 @@
 
 static struct memtier_tier *current_tier;
 
+typedef struct fs_dax_registy {
```
Comment (PatKamin) [inline tiering/ctl.c:20]: nit: registry

File: tiering/ctl.c [hunk 2]
```diff
@@ -17,13 +17,42 @@
...
 typedef struct ctl_tier_cfg {
     char *kind_name;
     char *pmem_path;
     size_t pmem_size;
     unsigned ratio_value;
 } ctl_tier_cfg;
 
+static struct fs_dax_registy fs_dax_reg_g;
+
+static int ctl_add_pmem_to_fs_dax_reg(memkind_t kind)
```
Comment (PatKamin) [inline tiering/ctl.c:34]: add_pmem_kind

File: tiering/ctl.c [hunk 3]
```diff
@@ -17,13 +17,42 @@
...
+{
+    memkind_t *new_kinds =
+        memkind_realloc(MEMKIND_DEFAULT, fs_dax_reg_g.kinds,
+                        sizeof(fs_dax_reg_g.kinds) * (fs_dax_reg_g.size + 1));
+    if (!new_kinds)
+        return -1;
+    fs_dax_reg_g.kinds = new_kinds;
+    fs_dax_reg_g.kinds[fs_dax_reg_g.size] = kind;
+    fs_dax_reg_g.size += 1;
```
Comment (PatKamin) [inline tiering/ctl.c:43]: nit: ++

File: tiering/ctl.c [hunk 4]
```diff
@@ -256,8 +285,14 @@ static memkind_t ctl_get_kind(const ctl_tier_cfg *tier)
         kind = MEMKIND_DEFAULT;
         log_debug("kind_name: memkind_default");
     } else if (strcmp(tier->kind_name, "FS_DAX") == 0) {
-        // TODO handle FS_DAX here
-        log_debug("kind_name: %s", tier->kind_name);
+        memkind_create_pmem(tier->pmem_path, tier->pmem_size, &kind);
+        if (kind) {
+            // dummy allocation for initialize arena
```
Comment (PatKamin) [inline tiering/ctl.c:290]: Why initializing?

File: test/tiering_tests.py [hunk 5]
```diff
@@ -28,6 +29,16 @@ class Helper(object):
     # but don't test anything related to allocation policies
     default_policy = "POLICY_CIRCULAR"
 
+    def check_fs_dax_support(self):
+        fs_dax_path = os.environ.get('PMEM_PATH', '/tmp')
+        if fs_dax_path[-1] == '/':
```
Comment (PatKamin) [inline test/tiering_tests.py:34]: Try rstrip('/')

File: test/tiering_tests.py [hunk 6]
```diff
@@ -292,6 +303,14 @@ def test_FSDAX_wrong_size(self, pmem_size):
         assert output[0] == self.log_error_prefix + \
             "Failed to parse pmem size: " + pmem_size, "Wrong message"
 
+    def test_FSDAX_check_only_fs_dax(self):
+        pmem_path = os.environ.get('PMEM_PATH')
+        if not self.check_fs_dax_support():
+            pytest.skip("Missing FS DAX placed on" + pmem_path)
```
Comment (PatKamin) [inline test/tiering_tests.py:309]: mounted on

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 3 of 3 files at r8, 2 of 2 files at r9, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/596)* status: all files reviewed, 6 unresolved discussions (waiting on @michalbiesek)

**Review by bratpiorka — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/596#-:-MWdeFTz9_dKKLmV8x6l:bnfp4nl)* status: all files reviewed, 6 unresolved discussions (waiting on @michalbiesek)

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/596)* status: all files reviewed, 4 unresolved discussions (waiting on @PatKamin)

___
*[test/tiering_tests.py, line 34 at r6](https://reviewable.io/reviews/memkind/memkind/596#-MWZGprHPXB3zXIStqZd-r6-34:-MWddLQh77PA0M0IQNZc:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/66d23d75162ad78465a71995fc537ffcfbaffbfc/test/tiering_tests.py#L34)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

Try rstrip('/')
</blockquote></details>

Done.
___
*[test/tiering_tests.py, line 235 at r6](https://reviewable.io/reviews/memkind/memkind/596#-MWdXccX4EeDUE66ycRH:-MWddMPFCXPkHpdn1hwt:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/66d23d75162ad78465a71995fc537ffcfbaffbfc/test/tiering_tests.py#L235)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

with this change retcode should be validated
</blockquote></details>

Done.
___
*[test/tiering_tests.py, line 252 at r6](https://reviewable.io/reviews/memkind/memkind/596#-MWdXlNvD3dfaZ7rLi6n:-MWdcgXH3J9WsyaK7Xf5:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/66d23d75162ad78465a71995fc537ffcfbaffbfc/test/tiering_tests.py#L252)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.
___
*[test/tiering_tests.py, line 309 at r6](https://reviewable.io/reviews/memkind/memkind/596#-MWZGprHPXB3zXIStqZd-r6-309:-MWdbbiY5r4flqXaYDZs:bb6un1j) ([raw file](https://github.com/memkind/memkind/blob/66d23d75162ad78465a71995fc537ffcfbaffbfc/test/tiering_tests.py#L309)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

mounted on
</blockquote></details>

Done
___
*[tiering/ctl.c, line 20 at r5](https://reviewable.io/reviews/memkind/memkind/596#-MWZGprGKNhJXUqxbqMq-r5-20:-MWdagwA-iVR8Uth51JM:bb6un1j) ([raw file](https://github.com/memkind/memkind/blob/3ef3aad49ad6ff2dae8a4e5d50185d0daa3cccf1/tiering/ctl.c#L20)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

nit: registry
</blockquote></details>

Done
___
*[tiering/ctl.c, line 20 at r5](https://reviewable.io/reviews/memkind/memkind/596#-MWdWzEl16p7fLbm6veI:-MWdagXp0cnDvj_biaSV:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/3ef3aad49ad6ff2dae8a4e5d50185d0daa3cccf1/tiering/ctl.c#L20)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

registy -> registry
</blockquote></details>

Done.
___
*[tiering/ctl.c, line 34 at r5](https://reviewable.io/reviews/memkind/memkind/596#-MWZGprGKNhJXUqxbqMq-r5-34:-MWdbD1b1OAAtl2sgJbs:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/3ef3aad49ad6ff2dae8a4e5d50185d0daa3cccf1/tiering/ctl.c#L34)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

add_pmem_kind
</blockquote></details>

Done.
___
*[tiering/ctl.c, line 43 at r5](https://reviewable.io/reviews/memkind/memkind/596#-MWZGprGKNhJXUqxbqMq-r5-43:-MWdbC-M0tJM9dTQ5U4E:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/3ef3aad49ad6ff2dae8a4e5d50185d0daa3cccf1/tiering/ctl.c#L43)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

nit: ++
</blockquote></details>

Done.
___
*[tiering/ctl.c, line 290 at r5](https://reviewable.io/reviews/memkind/memkind/596#-MWZGprGKNhJXUqxbqMq-r5-290:-MWdbEBQ-EpfoJFnlYVJ:b8kqfq) ([raw file](https://github.com/memkind/memkind/blob/3ef3aad49ad6ff2dae8a4e5d50185d0daa3cccf1/tiering/ctl.c#L290)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

Why initializing?
</blockquote></details>

Removed - this is needed with other kinds - see that there is additional action on the first malloc ( check what happened if use MEMKIND_REGULAR)

**Review by PatKamin — COMMENTED** [review]
> Reviewed 1 of 3 files at r8, 1 of 2 files at r9, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/596)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @michalbiesek)

**Pull request merged**